spark/src/main/scala/geotrellis/spark/io/s3/RasterRDDReader.scala

```diff
@@ -38,36 +40,41 @@ abstract class RasterRDDReader[K: ClassTag] extends LazyLogging {
     logger.debug(s"Loading $layerId from $dir")
 
     val ranges = setFilters(filterSet, keyBounds, keyIndex)
-    val bins = ballancedBin(ranges, sc.defaultParallelism)
-    logger.debug(s"Created ${ranges.length} ranges, binned into ${bins.length} bins")
+    val bins = balancedBin(ranges, numPartitions)
+    logger.info(s"Created ${ranges.length} ranges, binned into ${bins.length} bins")
 
     val bcClient = sc.broadcast(s3client)
     val bcFilterSet = sc.broadcast(filterSet)
-    val itp = indexToPath
-    val pti = pathToIndex
+    val maxWidth = maxIndexWidth(keyIndex.toIndex(keyBounds.maxKey))
+
+    // Broadcast the functions that translate from a path to an index and back
+    val bcFuncs = sc.broadcast((pathToIndex, { (index: Long) => indexToPath(index, maxWidth) }))
 
     val rdd =
-      sc.parallelize(bins)
-      .mapPartitions{ rangeList =>
-        val s3client: S3Client = bcClient.value.apply
-        val filterSet: FilterSet[K] = bcFilterSet.value
-
-        rangeList
-          .flatMap( ranges => ranges)
-          .flatMap{ range =>
-            listKeys(s3client, bucket, dir, range, itp, pti)            
-              .map{ path => 
-                val is = s3client.getObject(bucket, path).getObjectContent
-                KryoSerializer.deserializeStream[(K, Tile)](is)
-              }
-              .filter{ row => filterSet.includeKey(row._1) }
-          }
-      }
+      sc
+        .parallelize(bins, bins.size)
```

Comment:
I wonder if this is a place where we could take advantage of the built-in s3 and binary records support for spark -- maybe something like

```scala
sc.binaryRecords("s3a://file1,s3a://file2...", numPartitions)
                            .map{ case (filename, bytes) => KryoSerializer.deserializeStream[(K, Tile)](bytes) }
                            .filter{ row => filterSet.includeKey(row._1) }
```